paddle/fluid/operators/set_value_op.h

```diff
@@ -0,0 +1,198 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include <utility>
+#include "paddle/fluid/framework/eigen.h"
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/assign_value_op.h"
+#include "paddle/fluid/operators/elementwise/elementwise_op_function.h"
+#include "paddle/fluid/platform/enforce.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+inline std::string GetValueName(framework::proto::VarType::Type data_type) {
+  std::string value_name;
+  switch (data_type) {
+    case framework::proto::VarType::INT32:
+      value_name = "int32_values";
+      break;
+    case framework::proto::VarType::INT64:
+      value_name = "int64_values";
+      break;
+    case framework::proto::VarType::FP32:
+      value_name = "fp32_values";
+      break;
+    case framework::proto::VarType::BOOL:
+      value_name = "bool_values";
+      break;
+    default:
+      PADDLE_THROW(platform::errors::Unimplemented(
+          "Unsupported data type(code %d) for SetValue operator, only "
+          "supports bool, int32, float32 and int64.",
+          data_type));
+  }
+  return value_name;
+}
+
+inline framework::DDim GetSliceDims(const framework::DDim in_dims,
+                                    const std::vector<int64_t> axes,
+                                    const std::vector<int64_t> starts,
+                                    const std::vector<int64_t> ends) {
+  framework::DDim slice_dims(in_dims);
+
+  for (size_t i = 0; i < axes.size(); ++i) {
+    int64_t axis = axes[i];
+    int64_t dim_value = in_dims[axis];
+
+    int64_t start = starts[i] < 0 ? (starts[i] + dim_value) : starts[i];
+    int64_t end = ends[i] < 0 ? (ends[i] + dim_value) : ends[i];
+    start = std::max(start, static_cast<int64_t>(0));
+    end = std::min(end, dim_value);
+
+    PADDLE_ENFORCE_GT(end, start, platform::errors::InvalidArgument(
+                                      "end should greater than start, but "
+                                      "received end = %d, start = %d",
+                                      end, start));
+    slice_dims[axis] = end - start;
+  }
+  return slice_dims;
+}
+
+template <typename DeviceContext, typename T>
+class SetValueKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const {
+    const int rank = ctx.Output<framework::LoDTensor>("Out")->dims().size();
+    switch (rank) {
```

Comment:
So amazing, I just knew it from this PR. This code is ugly. I can LGTM here now since I found C++ actually has to make template integer as constant as what you said, but we had better have alternative writing in the future.

Reply:
Thanks.